Add Multiple Tower Support in Condenser Entering Temp Reset Setpoint Managers

Myoldmopar opened the pull request:
This PR will ultimately be the implementation of allowing multiple towers in the CondenserEnteringTemp reset setpoint managers.  As of right now, I have done some restructuring of the SetPointManager code to lean it toward OO patterns and some other minor cleanup.  
- [x] Verify initial changes have no-diffs
- [x] Provide NFP to team
- [x] Provide design document
- [x] Draft code push
- [x] Code review
- [x] Final code submission
- [x] Documentation changes

Myoldmopar committed:
Modify changelog script to avoid unrelated PRs

Myoldmopar committed:
Remove unused member constructors from setpoint managers

Myoldmopar committed:
Restructure Setpoint manager calc functions into member functions

Myoldmopar committed:
Add lambda support to changelog script for @amirroth and @mbadams5

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 94006714-MultipleTowers

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 94006714-MultipleTowers

Myoldmopar committed:
Minor cleanup in setpoint manager; add design doc

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 94006714-MultipleTowers

Myoldmopar committed:
Add new example files, stub out changes to start allowing multiple towers

Myoldmopar committed:
Update design doc, add figure

Myoldmopar committed:
Update design doc and idf;
Multiple towers appear to work for non-ideal reset

Myoldmopar committed:
Update plot

Myoldmopar committed:
Update docs for multiple towers

Myoldmopar commented:
Hmmmm, so I'm seeing a potential issue in the ideal condenser reset manager.  I'll try to tell the story with lines of code here.

During init, we search out for the condenser loop that contains the setpoint node defined in the setpoint manager object.  No problem.  Once we've identified that loop, we then loop over the demand side branches [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L3683) with `BranchNum` and components [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L3684) with `CompNum`.  We check the `TypeOf_Num` for that component and if it is a chiller, we start doing additional mining to pick up the chiller and pump branch, comp.  But within the same loop, we then set the following variables:

```
IdealCondEntSetPtMgr( SetPtMgrNum ).TypeNum = TypeNum;
IdealCondEntSetPtMgr( SetPtMgrNum ).CondLoopNum = LoopNum;
IdealCondEntSetPtMgr( SetPtMgrNum ).TowerNum = CompNum;
IdealCondEntSetPtMgr( SetPtMgrNum ).CondBranchNum = BranchNum;
```

At this point, `CompNum` is still assigned to the one checked previously that was used to identify the chiller component.  Yet it is being assigned to the `TowerNum`.  This is probably passing under the radar because there is usually one component per branch.  But the bigger point is that `BranchNum` has the same effect, and is being stored in `CondBranchNum`.  [Later](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L6657) in the setpoint manager operation, this variable is then assigned to a local variable named `TowerBranchNum`, and used to look up the meters.  This appears to be blatantly wrong.  As I see it, the tower energy is not being used in searching for an optimal energy use, and the chiller branch num is being double counted.

Can someone spare a minute to verify I'm not missing something obvious?  Because if I'm right, this means this setpoint manager has been wrong from the [beginning](https://github.com/NREL/EnergyPlus/blame/develop/src/EnergyPlus/SetPointManager.cc#L3722).

FYI: @EnergyArchmage @mjwitte @rraustad @lgu1234

rraustad commented:
It looks to me as if the 3 variables together should point right to the 
chiller.

PlantLoop( LoopNum ).LoopSide( \* ).Branch( BranchNum ).Comp( CompNum )

So I am not quite sure what this line is supposed to do since it points 
to a chiller:
TowerBranchNum = IdealCondEntSetPtMgr( SetPtMgrNum ).CondBranchNum;

Is there another loop like this somewhere looking for a tower? If there 
is, it would point to the tower for that SetPtMgrNum

On 8/20/2015 1:34 PM, Edwin Lee wrote:

> Hmmmm, so I'm seeing a potential issue in the ideal condenser reset 
> manager. I'll try to tell the story with lines of code here.
> 
> During init, we search out for the condenser loop that contains the 
> setpoint node defined in the setpoint manager object. No problem. Once 
> we've identified that loop, we then loop over the demand side branches 
> here 
> https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L3683 
> with |BranchNum| and components here 
> https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L3684 
> with |CompNum|. We check the |TypeOf_Num| for that component and if it 
> is a chiller, we start doing additional mining to pick up the chiller 
> and pump branch, comp. But within the same loop, we then set the 
> following variables:
> 
> |IdealCondEntSetPtMgr( SetPtMgrNum ).TypeNum = TypeNum; 
> IdealCondEntSetPtMgr( SetPtMgrNum ).CondLoopNum = LoopNum; 
> IdealCondEntSetPtMgr( SetPtMgrNum ).TowerNum = CompNum; 
> IdealCondEntSetPtMgr( SetPtMgrNum ).CondBranchNum = BranchNum; |
> 
> At this point, |CompNum| is still assigned to the one checked 
> previously that was used to identify the chiller component. Yet it is 
> being assigned to the |TowerNum|. This is probably passing under the 
> radar because there is usually one component per branch. But the 
> bigger point is that |BranchNum| has the same effect, and is being 
> stored in |CondBranchNum|. Later 
> https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L6657 
> in the setpoint manager operation, this variable is then assigned to a 
> local variable named |TowerBranchNum|, and used to look up the meters. 
> This appears to be blatantly wrong. As I see it, the tower energy is 
> not being used in searching for an optimal energy use, and the chiller 
> branch num is being double counted.
> 
> Can someone spare a minute to verify I'm not missing something 
> obvious? Because if I'm right, this means this setpoint manager has 
> been wrong from the beginning 
> https://github.com/NREL/EnergyPlus/blame/develop/src/EnergyPlus/SetPointManager.cc#L3722.
> 
> FYI: @EnergyArchmage https://github.com/EnergyArchmage @mjwitte 
> https://github.com/mjwitte @rraustad https://github.com/rraustad 
> @lgu1234 https://github.com/lgu1234
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/NREL/EnergyPlus/pull/5088#issuecomment-133088107.

## 

Richard Raustad
Florida Solar Energy Center
1679 Clearlake Road
Cocoa, FL 32922
Ph: (321)638-1454
http://www.fsec.ucf.edu/en/

Program Director
Electric Vehicle Transportation Center
http://evtc.fsec.ucf.edu/

Myoldmopar commented:
Yeah, that's what I'm seeing too @rraustad.  That line I showed above is the only assignment of the `CondBranchNum` variable, so I think that is a blatant bug.  Crap, now I'm torn as to whether to include the bug fix here or do it separate.  Alright, well I'll fix it either way.

rraustad commented:
This should be IdealCondEntSetPtMgr().CompNum (or since it's a chiller, ChillerNum):

IdealCondEntSetPtMgr( SetPtMgrNum ).TowerNum = CompNum;

The chiller index is stored here:
ChillerNum = IdealCondEntSetPtMgr( SetPtMgrNum ).ChillerIndexPlantSide;

And since the for loops only test for chillers, why are these even there:

TowerLoopNum = IdealCondEntSetPtMgr( SetPtMgrNum ).CondLoopNum;
TowerBranchNum = IdealCondEntSetPtMgr( SetPtMgrNum ).CondBranchNum;
TowerNum = IdealCondEntSetPtMgr( SetPtMgrNum ).TowerNum;

Oh, I see, the tower loop is just above the chiller loop. The only problem is they never stored the data (no assignments).

rraustad commented:
You should probably have:
IdealCondEntSetPtMgr().ChillerCompNum 
IdealCondEntSetPtMgr().TowerCompNum 

since there can be a SPM for each of these, or one for one and not another. That way you'd know what component the SPM controlled.

Myoldmopar commented:
@rraustad 

> since there can be a SPM for each of these

I'm not sure I get what you mean here.  This is a pretty high level setpoint manager that sets the condenser loop setpoint by querying the current state of the chiller, tower, and both pumps.  

Ultimately we will need some representation of the following:
- Scalars: ChillerBranchNum, ChillerCompNum
- Vectors: TowerBranchNums, TowerCompNums (since now allowing multiple towers, better ways to store multiple of these than separate arrays...)

Myoldmopar committed:
Modifications to allow multiple towers;
builds, runs single and multiple tower files for both regular and ideal types;
surely still needs work and agreement on the design, but at least many questions are answered now

Myoldmopar commented on doc/tools/create_changelog.py:
```diff
@@ -68,6 +68,16 @@ def usage():
+    expected_good_num = lambda n : int(pr_num) < 1000
+    if expected_good_num(pr_num):
+        continue
+    
```
The above is an unrelated fix to clean up the changelog error handling; it doesn't actually affect the output of the changelog at this time.

Myoldmopar commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -6686,16 +6639,19 @@ namespace SetPointManager {
+				TowerFanEnergy = 0;
+				for ( int i = 1; i <= this->numTowers; i++ ) {
+					TowerFanEnergy += GetInternalVariableValue( this->ClTowerVarType( i ), this->ClTowerVarIndex( i ) );
+				}
```
Now loops over all towers on the loop to get the total energy use.  This loop is the core of this feature.

Myoldmopar commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -7076,45 +6992,30 @@ namespace SetPointManager {
+			GetMeteredVariables( TypeOfComp, NameOfComp, VarIndexes, VarTypes, IndexTypes, UnitsStrings, ResourceTypes, EndUses, Groups, Names, NumFound );
+			this->ClTowerVarType.push_back( VarTypes( 1 ) );
+			this->ClTowerVarIndex.push_back( VarIndexes( 1 ) );
+
```
This loop gets the var type and var index for multiple towers on the condenser loop.

Myoldmopar commented:
Well, if you don't count the new idfs I added, I am currently at a net of -781 lines of code with this PR.  I need to do a little more testing, and update the design doc with my current plans to complete it so I can put that out to review.  I know I have already done some coding, but I felt I needed to do that to determine the design.  [:egg:, :chicken:].first?

Myoldmopar committed:
Starting unit test for spm work

Myoldmopar committed:
Continued work toward final unit test

Myoldmopar committed:
Resolve merge conflict from TES setpoint work

Myoldmopar committed:
Move TES code to a member function to match my previous OO work

Myoldmopar committed:
Update unit test to actually do testing; works well now

Myoldmopar committed:
Restructured ideal cond setpoint calcs for easier unit testing;
Added unit test for new member function, runs fine locally;
This could wrap up this task

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 94006714-MultipleTowers

Myoldmopar committed:
Clean up some misc warnings from the new example files; output looks fine

Myoldmopar committed:
Update design doc with a small wrap up section

mjwitte commented:
@Myoldmopar So, why don't the results change for CoolingTower_VariableSpeed_IdealCondEntTempSetpoint_MultipleTowers?  Ran this file as an annual simulation before and after these changes and the ABPUS report is identical.

mjwitte commented:
Found that CoolingTower_VariableSpeed_IdealCondEntTempSetpoint_MultipleTowers test file didn't really have two towers.  Once this is updated, this can go in.

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 94006714-MultipleTowers

Myoldmopar committed:
Actually use two towers in the ideal multiple tower test file...

Myoldmopar committed:
Fix line ending issue

Myoldmopar committed:
Update idf for multiple towers; update design doc with discussion for multiple towers

Myoldmopar commented:
@mjwitte I updated the idf for multiple towers, pulled in develop, rebuilt, and re-ran.  I got good results.  I updated the [final section](https://github.com/NREL/EnergyPlus/blob/94006714-MultipleTowers/design/FY2015/CondenserResetWithMultipleTowers.md#modifications-and-testing-of-the-ideal-condenser-setpoint-manager) of the design doc with improved discussion.  I think this is ready to go in if you don't see any other issues.

Myoldmopar commented:
And for the EnergyPlus source itself, I still ended up with a net lines of code added = -990.  w00t.

mjwitte commented:
That didn't change anything.  For the ideal reset manager, the two-tower case still uses 2x heat rejection electricity compared to the single-tower case.

Myoldmopar commented:
OK, I'm about ready to wash my hands of this.  It's late in the game and I'm in the mood to fix bugs.  We can chat on the sizing call about whether to push forward or not.

mjwitte commented:
Once CI comes back with a couple more, I'll merge this and post a followup issue #5207
